tuf/testutils/swizzler.go

```diff
@@ -0,0 +1,551 @@
+	// in a separate loop because "targets/a" can't be serialized until "targets/a/b"
+	// has been added
+	for _, delgName := range delegatedRoles {
+		signedThing, err := tufRepo.SignTargets(delgName, data.DefaultExpires("targets"))
+		if err != nil {
+			return nil, err
+		}
+		metaBytes, err := json.MarshalCanonical(signedThing)
+		if err != nil {
+			return nil, err
+		}
+		meta[delgName] = metaBytes
+	}
+
+	// these need to be generated after the delegations are created and signed so
+	// the snapshot will have the delegation metadata
+	rs, tgs, ss, ts, err := Sign(tufRepo)
+	if err != nil {
+		return nil, err
+	}
+	rf, tgf, sf, tf, err := Serialize(rs, tgs, ss, ts)
+	if err != nil {
+		return nil, err
+	}
+
+	meta[data.CanonicalRootRole] = rf
+	meta[data.CanonicalSnapshotRole] = sf
+	meta[data.CanonicalTargetsRole] = tgf
+	meta[data.CanonicalTimestampRole] = tf
+
+	swizzler := MetadataSwizzler{
+		gun:           gun,
+		MetadataCache: store.NewMemoryStore(meta, nil),
+		CryptoService: cs,
+		Roles:         DefaultRoles,
+	}
+
+	return &swizzler, nil
+}
+
+// SetInvalidJSON corrupts metadata into something that is no longer valid JSON
+func (m *MetadataSwizzler) SetInvalidJSON(role string) error {
+	metaBytes, err := m.MetadataCache.GetMeta(role, maxSize)
+	if err != nil {
+		return err
+	}
+	return m.MetadataCache.SetMeta(role, metaBytes[5:])
+}
+
+// SetInvalidSigned corrupts the metadata into something that is valid JSON,
+// but not unmarshallable into signed JSON
+func (m *MetadataSwizzler) SetInvalidSigned(role string) error {
+	signedThing, err := signedFromStore(m.MetadataCache, role)
+	if err != nil {
+		return err
+	}
+	metaBytes, err := json.MarshalCanonical(map[string]interface{}{
+		"signed":     signedThing.Signed,
+		"signatures": "not list",
+	})
+	if err != nil {
+		return err
+	}
+	return m.MetadataCache.SetMeta(role, metaBytes)
+}
+
+// SetInvalidSignedMeta corrupts the metadata into something that is unmarshallable
+// as a Signed object, but not unmarshallable into a SignedMeta object
+func (m *MetadataSwizzler) SetInvalidSignedMeta(role string) error {
+	signedThing, err := signedFromStore(m.MetadataCache, role)
+	if err != nil {
+		return err
+	}
+
+	var unmarshalled map[string]interface{}
+	if err := json.Unmarshal(signedThing.Signed, &unmarshalled); err != nil {
+		return err
+	}
+
+	unmarshalled["_type"] = []string{"not a string"}
+	unmarshalled["version"] = "string not int"
+	unmarshalled["expires"] = "cannot be parsed as time"
+
+	metaBytes, err := json.MarshalCanonical(unmarshalled)
+	if err != nil {
+		return err
+	}
+	signedThing.Signed = json.RawMessage(metaBytes)
+
+	metaBytes, err = serializeMetadata(m.CryptoService, signedThing, role)
+	if err != nil {
+		return err
+	}
+	return m.MetadataCache.SetMeta(role, metaBytes)
+}
+
+// TODO: corrupt metadata in such a way that it can be unmarshalled as a
+// SignedMeta, but not as a SignedRoot or SignedTarget, etc. (Signed*)
```

Comment:
I think maybe the Dirty field doesn't get serialized as JSON?  I thought that was just something we use for accounting ourselves.